### PR TITLE
fix(mocha): correct timing for top-level it() tests

### PR DESCRIPTION
**Pull request by cbasitodx**
### What does this PR do?

This PR fixes a bug in the Mocha CI Visibility instrumentation triggered by test files that contain top-level `it()` calls with no wrapping `describe()` block. It follows [this](https://github.com/DataDog/dd-trace-js/pull/8437) PR.

## Bug: Suite span timing (all files start and end simultaneously)

**Root cause.** Suite spans for top-level-only files were created inside the root `'suite'` handler, which fires *before any test runs*. When two or more such files are run together, all their suite spans start at the same moment (root `'suite'` time) and finish at root `'suite end'` time (after all files have run). Suite durations therefore include time spent on other files, and spans overlap.

**What changes (lazy start).** Instead of calling `testSuiteStartCh.runStores` immediately in the root `'suite'` handler, the ctx is stored with a `_pendingRootStart: true` flag. A new `'test'` listener (registered before `getOnTestHandler`) fires `testSuiteStartCh` on the *first test event for that file*, matching the describe-based behaviour where the suite span opens just before the first test runs.

Here is a little high-level example:

```
// Before fix — all suites start at t=0 (root 'suite' time):
file-a suite: [t=0 ──────────────────────────── t=500ms]
file-b suite: [t=0 ──────────────────────────── t=500ms]

// After fix — each suite starts when its own file begins:
file-a suite: [t=10ms ────── t=250ms]
file-b suite:                [t=250ms ── t=500ms]
```

**What changes (per-file finish).** `_ddIsFinalAttempt` is set synchronously on the test object in `getOnTestEndHandler` and `getOnHookEndHandler` (covering tests with `afterEach` hooks) whenever `finalStatus !== undefined`. This flag is reliable across all retry mechanisms: plain tests (always set), ATR (set on the last `this.retries(N)` attempt), EFD (set on the last EFD clone), and ATF (set on the last attempt-to-fix run). [(mocha final status PR)](https://github.com/DataDog/dd-trace-js/pull/7844)

A countdown map `rootPendingCountByFile` is initialised from `rootTestsByFile` at `'suite'` time. Two new listeners (`'test end'` and `'hook end'`) decrement the count per file on each `_ddIsFinalAttempt`. When the count hits zero, `finishRootSuiteForFile` fires: it publishes code coverage (scoped to that file's tests only. Reset happens immediately after, so the next file starts with a clean slate) and calls `testSuiteFinishCh`.

The all-pending edge case (no `'test'` event fires at all) is handled in root `'suite end'`: if `_pendingRootStart` is still set, the suite is started and immediately finished with `status: skip`.

**Coverage correctness.** The old root `'suite end'` loop published coverage after *all* root tests had run, so the first file got the combined global coverage and subsequent files got an empty object. The per-file finish approach resets after each file's last test, giving each file only its own coverage.

### Additional Notes

- Added corresponding integration tests

### CODE REVIEW

File: packages/datadog-instrumentations/src/mocha/main.js [hunk 1]
```diff
@@ -469,26 +469,124 @@ addHook({
...
+
     this.on('retry', getOnTestRetryHandler(config))
 
     // If the hook passes, 'hook end' will be emitted. Otherwise, 'fail' will be emitted
     this.on('hook end', getOnHookEndHandler(config))
 
+    this.on('hook end', function (hook) {
+      const test = hook.ctx?.currentTest
+      if (!test?._ddIsFinalAttempt || !rootPendingCountByFile.has(test.file)) return
+      finishRootSuiteForFile(test.file)
```
[the remark this answers is not in the document]
Comment (cbasitodx) [reply packages/datadog-instrumentations/src/mocha/main.js]: Nice catch

**Comment by cbasitodx** [comment]
@codex review

File: packages/datadog-instrumentations/src/mocha/main.js [hunk 2]
```diff
@@ -469,26 +469,131 @@ addHook({
...
+      }
+    })
+
     this.on('test', getOnTestHandler(true))
 
     this.on('test end', getOnTestEndHandler(config))
 
+    this.on('test end', function (test) {
+      if (!test._ddIsFinalAttempt || !rootPendingCountByFile.has(test.file)) return
+      finishRootSuiteForFile(test.file)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/mocha/main.js]: I think we can tackle this in a separate PR

File: integration-tests/mocha/mocha.spec.js [hunk 3]
```diff
@@ -1263,6 +1263,170 @@ describe(`mocha@${MOCHA_VERSION}`, function () {
 
           await Promise.all([eventsPromise, once(childProcess, 'exit')])
         })
+
+        it.only('reports a suite event when a beforeEach hook fails on top-level tests', async () => {
```
Comment (juan-fernandez) [inline integration-tests/mocha/mocha.spec.js]:
```suggestion
        it('reports a suite event when a beforeEach hook fails on top-level tests', async () => {
```

**Review by juan-fernandez — CHANGES_REQUESTED** [review]
let's remove the `it.only`. Otherwise LGTM

**Comment by juan-fernandez** [comment]
## E2E Test Report: SUCCESS

**Tested by:** Shepherd Agent (autonomous QA for Datadog Test Optimization)

### Test Environment

- **Method:** Local testing with Shepherd `express` playground using Mocha and mockdog
- **PR:** https://github.com/DataDog/dd-trace-js/pull/8480
- **Branch tested:** `sebas_conde/fix/no_suite_bug_addition`
- **Revision tested:** `b985bb1d42a664a63fc28dce3df14095f3560745`
- **Mockdog run:** `20260518-114506`
- **Command:** `./bin/crook run express --dep dd-trace-js=sebas_conde/fix/no_suite_bug_addition --scenario coverage-enabled --debug -c single -f test/shepherd-pr-8480-top-level-a.js -- test/shepherd-pr-8480-top-level-b.js`

### Results

| Check | Result |
| --- | --- |
| Top-level-only Mocha files run in one process | PASS |
| Test spans emitted | PASS: 4 tests, all passed |
| Suite spans emitted per file | PASS: 2 suites, all passed |
| Suite timing corrected | PASS: file B suite starts after file A suite ends, no root-wide overlap |
| Suite source files | PASS: `test/shepherd-pr-8480-top-level-a.js`, `test/shepherd-pr-8480-top-level-b.js` |
| Code coverage enabled on session | PASS: `test.code_coverage.enabled=true` |
| Coverage publication | PASS: 1 `/api/v2/citestcov` request, 2 coverage entries, 4 files covered |
| Coverage scoping | PASS: file A coverage contained `lib/application.js` + file A; file B coverage contained `lib/response.js` + file B |
| Telemetry | PASS: `code_coverage_started=2`, `code_coverage_finished=2`, `code_coverage_errors=0`, `code_coverage_payloads=1` |

### Verification

Local mockdog backend verification passed. The run recorded `17` CI Test Cycle events: `4` tests, `2` suites, `1` module, `1` session. It also recorded `2` code coverage entries in one `/api/v2/citestcov` payload.

Datadog UI verification was not performed because this E2E run targeted local mockdog rather than a real Datadog site.

### Issues Found

No PR-blocking issues found.

Harness note: a plain Mocha run does not create `global.__coverage__`, so the final focused test seeded a minimal Istanbul coverage map in temporary playground helper files to verify the PR coverage publication path. With seeded coverage, mockdog received separate coverage entries for each top-level-only test file.

### Test Methodology

1. Selected the `express` playground because it uses Mocha and is the Shepherd JS playground for dd-trace-js Mocha plugin testing.
2. Created two temporary top-level-only Mocha files, each containing top-level `it()` calls and no wrapping `describe()`.
3. Ran both files in a single Mocha process with the PR branch injected via `--dep dd-trace-js=sebas_conde/fix/no_suite_bug_addition`.
4. Enabled the `coverage-enabled` mockdog scenario and `--debug` logging.
5. Verified parsed mockdog spans with `bin/query-spans` and inspected the raw `/api/v2/citestcov` payload.

---

_This E2E test was performed by Shepherd - autonomous QA agent for Datadog Test Optimization._

**Comment by cbasitodx** [comment]
@codex review

File: packages/datadog-instrumentations/src/mocha/main.js [hunk 4]
```diff
@@ -473,22 +496,97 @@ addHook({
...
+      }
+    })
+
     this.on('test', getOnTestHandler(true))
 
-    this.on('test end', getOnTestEndHandler(config))
+    this.on('test end', getOnTestEndHandler(config, function (test) {
+      if (!test._ddIsFinalAttempt || !rootPendingCountByFile.has(test.file)) return
+      updateRootTestForFinalAttempt(test)
+      finishRootSuiteForFile(test.file)
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/mocha/main.js]: let's tackle this in a separate PR

File: packages/datadog-instrumentations/src/mocha/main.js [hunk 5]
```diff
@@ -536,42 +640,36 @@ addHook({
...
+          if (!ctx) continue
+          if (ctx._pendingRootStart) {
+            ctx._pendingRootStart = false
+            testSuiteStartCh.runStores(ctx, () => {})
+            testSuiteFinishCh.publish({ status: 'skip', ...ctx.currentStore }, () => {})
+          } else if (rootPendingCountByFile.has(test.file)) {
+            const rootTests = rootTestsByFile.get(test.file) || []
+            const status = haveRootTestsFinished(rootTests) ? getRootSuiteStatus(rootTests) : 'fail'
+            rootPendingCountByFile.delete(test.file)
             testSuiteFinishCh.publish({ status, ...ctx.currentStore }, () => {})
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/mocha/main.js]: let's tackle this in a separate PR

**Review by juan-fernandez — APPROVED** [review]
